tests/test_s3urlcache.py

```diff
@@ -0,0 +1,196 @@
+        self.assertEqual(url_content, original_data)
+        self.assertTrue(log_monitor.output[0].endswith(f"{url} not found in cache. Adding it to "
+                                                       f"{self.test_bucket} with key {url_key}."))
+
+    def test_retrieve_from_cache(self):
+        """Stored URL contents is retrieved from S3, when a cached url is requested"""
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        with self.assertLogs(logger, 'INFO') as log_monitor:
+            url_content = self.cache.resolve(url)
+            cached_content = self.cache.resolve(url)
+        self.assertEqual(len(log_monitor.output), 1)
+        self.assertEqual(url_content, cached_content)
+
+    def test_bad_url(self):
+        bad_urls = ['', '//', 'http://?']
+        self.urls_to_cleanup.update(bad_urls)
+
+        for url in bad_urls:
+            with self.subTest(bad_url=url):
+                with self.assertRaises(requests.RequestException):
+                    self.cache.resolve(url)
+
+    def test_url_SizeLimitError(self):
+        """Exception returned when URL content size is greater than max_size."""
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        self.cache.max_size = 1
+        with self.assertRaises(SizeLimitError) as ex:
+            self.cache.resolve(url)
+        self.assertEqual(ex.exception.args[0], f"{url} not cached. The URL's contents have exceeded "
+                                               f"{self.cache.max_size} bytes.")
+
+    def test_content_size_0(self):
+        url = f"{HTTPInfo.url}/0"
+        self.urls_to_cleanup.add(url)
+        url_content = self.cache.resolve(url)
+        self.assertEqual(len(url_content), 0)
+
+    def test_chunked_content(self):
+        size = KiB * 10
+        url = f"{HTTPInfo.url}/{size}"
+        self.urls_to_cleanup.add(url)
+        self.cache.max_size = 1 * MB
+        self.cache.chunk_size = 1 * KiB
+        url_content = self.cache.resolve(url)
+        original_data = randomdata[:size]
+        self.assertEqual(url_content, original_data)
+
+    def test_stored_url_metadata(self):
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        url_key = S3UrlCache._url_to_key(url)
+        self.cache.resolve(url)
+
+        with self.subTest("check dss_cached_url"):
+            metadata = self.cache._reverse_key_lookup(url_key)
+            self.assertEqual(metadata, url)
+
+        with self.subTest("check content_type"):
+            contentType = self.blobstore.get_content_type(self.test_bucket, url_key)
+            self.assertEqual(contentType, "application/octet-stream")
+
+    def test_evict(self):
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        url_key = S3UrlCache._url_to_key(url)
+
+        with self.assertLogs(logger, "INFO") as log_monitor:
+            # Check when deleting url that is not cached.
```

Comment:
The proof that the URL is not cached is a bit weak. It would be a stronger proof of `evict`'s idempotence to call it twice in a row, without `resolve` in between. Also, log messages don't prove much, only that the log message was emitted, but not that the action was actually performed.

IOW, I would write this test as follows

```
resolve()
assert present
evict()
assert missing
evict()
assert missing
```

For the asserts I would add a `S3UrlCache.contains(url)` method that does a cheap HEAD request against the key. That method might come in handy outside of tests, too.

Reply:
fixed